### PR TITLE
Fix #2116

### LINKED ISSUE #2116: Support glob/grep in `ls`

### PR DESCRIPTION
**Pull request by alvarofe**
Added glob support for ls.

https://asciinema.org/a/18782

Anything that could be wrong just tell me ;)

### CODE REVIEW

**Comment by alvarofe** [comment]
Updated

File: libr/core/syscmd.c [hunk 1]
```diff
@@ -104,6 +107,23 @@ R_API void r_core_syscmd_ls(const char *input) {
 			}
 		} else path = input+2;
 	}
+	if (!r_file_is_directory (path)){
+		p = strrchr(path, '/');
+		if (p){
+			off = p - path;
+			d = (char *) calloc (1, off);
+			if (!d){
+				free (pattern);
```
Comment (alvarofe) [inline libr/core/syscmd.c]: Here I'm freeing pattern without allocate previously.

**Comment by radare** [comment]
Thanks! it would be good to write some tests to make r_str_glob() better and consistent. @maijin ?

**Comment by Maijin** [comment]
@radare I can make "ls" tests is that what you mean ?

**Comment by radare** [comment]
yes, but mainly for testing different glob strings. so we need a custom dir with few files with different extensions and such

**Comment by alvarofe** [comment]
ls ~/dir doesn't work since ~ is filtered for the grep functionality, isn't?

**Comment by radare** [comment]
quote the whole command to use this special char

“ls ~/dir” should work

and iirc also ls ${HOME}/ should be working

**Comment by alvarofe** [comment]
I will check that. @Maijin if after make the tests something is broken tell me ;)

**Comment by alvarofe** [comment]
none was working in the previous branch but the quoting is not filtered. I will update it to handle it.

In the case of ${HOME}/ r2 doesn't substitute the variable. I will look at it as well.

**Pull request closed without merging**